crates/sui-indexer/migrations/2023-03-01-193733_owner/up.sql

```diff
@@ -1,36 +1,133 @@
-    change_type            owner_change_type NOT NULL,
-    owner_type             owner_type        NOT NULL,
-    owner                  VARCHAR(66),
-    initial_shared_version BIGINT,
-    object_digest          VARCHAR           NOT NULL,
-    object_type            VARCHAR,
-    CONSTRAINT owner_changes_pk PRIMARY KEY (epoch, object_id, version)
-) PARTITION BY RANGE (epoch);
-CREATE INDEX owner_changes_owner_index ON owner_changes (owner_type, owner);
-CREATE TABLE owner_changes_partition_0 PARTITION OF owner_changes FOR VALUES FROM (0) TO (1);
+    epoch                  BIGINT       NOT NULL,
+    checkpoint             BIGINT       NOT NULL,
+    object_id              address PRIMARY KEY,
+    version                BIGINT       NOT NULL,
+    object_digest          base58digest NOT NULL,
+    owner_type             owner_type   NOT NULL,
+    owner_address          address,
+    initial_shared_version BIGINT
+);
+CREATE INDEX owner_epoch_index ON owner (epoch);
+CREATE INDEX owner_owner_index ON owner (owner_type, owner_address);
+
+-- store proc for updating owner table from objects table
+CREATE OR REPLACE FUNCTION owner_modified_func() RETURNS TRIGGER AS
+$body$
+BEGIN
+    IF (TG_OP = 'UPDATE') THEN
+        UPDATE owner
+        SET epoch                  = NEW.epoch,
+            checkpoint             = NEW.checkpoint,
+            version                = NEW.version,
+            object_digest          = NEW.object_digest,
+            owner_type             = NEW.owner_type,
+            owner_address          = NEW.owner_address,
+            initial_shared_version = NEW.initial_shared_version
+        WHERE object_id = NEW.object_id;
+        RETURN NEW;
+    ELSIF (TG_OP = 'DELETE') THEN
+        DELETE FROM owner WHERE object_id == OLD.object_id;
+        RETURN OLD;
+    ELSIF (TG_OP = 'INSERT') THEN
+        INSERT INTO owner
+        VALUES (NEW.epoch, NEW.checkpoint, NEW.object_id, NEW.version, NEW.object_digest, NEW.owner_type,
+                NEW.owner_address,
+                NEW.initial_shared_version);
+        RETURN NEW;
+    ELSE
+        RAISE WARNING '[OWNER_MODIFIED_FUNC] - Other action occurred: %, at %',TG_OP,NOW();
+        RETURN NULL;
+    END IF;
 
-CREATE TABLE owner_index
+EXCEPTION
+    WHEN data_exception THEN
+        RAISE WARNING '[OWNER_MODIFIED_FUNC] - UDF ERROR [DATA EXCEPTION] - SQLSTATE: %, SQLERRM: %',SQLSTATE,SQLERRM;
+        RETURN NULL;
+    WHEN unique_violation THEN
+        RAISE WARNING '[OWNER_MODIFIED_FUNC] - UDF ERROR [UNIQUE] - SQLSTATE: %, SQLERRM: %',SQLSTATE,SQLERRM;
+        RETURN NULL;
+    WHEN OTHERS THEN
+        RAISE WARNING '[OWNER_MODIFIED_FUNC] - UDF ERROR [OTHER] - SQLSTATE: %, SQLERRM: %',SQLSTATE,SQLERRM;
+        RETURN NULL;
+END;
+$body$
+    LANGUAGE plpgsql;
+
+CREATE TRIGGER owner
+    AFTER INSERT OR UPDATE OR DELETE
+    ON objects
+    FOR EACH ROW
+EXECUTE PROCEDURE owner_modified_func();
+
+CREATE TABLE owner_history
```

Comment:
Do you mind giving an example about how a query looks like for the `owner_history` table? How does the caller know which partition/epoch to fetch?

Reply:
thanks to this not so noob question, make me realise this PR was missing the deleted objects handling. I have since added that back so the owner_history query will now.

to query the "current" object ownership, we can simply query the `owner` table, which contain the latest owner index.
e.g.
```
SELECT *
FROM owner
WHERE owner_type = 'address_owner'
  AND owner_address =
      '0xb6a6dc32eb5833b4ef95407f5dd4a3abfd0042dfc619091bfc7335fb2864f41a'
  AND object_status NOT IN ('deleted', 'wrapped', 'unwrapped_then_deleted');
```

to query the owner_history table, it is a bit more involved:
```
SELECT *
FROM (SELECT *,
             RANK() OVER (PARTITION BY object_id ORDER BY version DESC) version_rank
      FROM owner_history
      WHERE ((owner_type = 'address_owner' AND
              owner_address = '0xb6a6dc32eb5833b4ef95407f5dd4a3abfd0042dfc619091bfc7335fb2864f41a')
          OR (old_owner_type = 'address_owner' AND
              old_owner_address = '0xb6a6dc32eb5833b4ef95407f5dd4a3abfd0042dfc619091bfc7335fb2864f41a'))
        AND checkpoint <= 20) AS diff
WHERE version_rank = 1
  AND owner_address = '0xb6a6dc32eb5833b4ef95407f5dd4a3abfd0042dfc619091bfc7335fb2864f41a';
```
this query select all owner changes (gain and loss ownership) from the history table for a address before checkpoint 20, and then for each object in the subquery, select the highest version number(using `RANK() OVER`), and then select the one belongs to the address. 

in simple terms this query will give you all object owned by that address at a checkpoint.

This is not easy to query so I have created a DB function to make it easier:
```
SELECT *
FROM object_owned_at_checkpoint(20, 'address_owner',
                                '0xb6a6dc32eb5833b4ef95407f5dd4a3abfd0042dfc619091bfc7335fb2864f41a');
```